proxy-blocker/background/proxy-handler.js

```diff
@@ -1,55 +1,51 @@
-// Location of the proxy script, relative to manifest.json
-const proxyScriptURL = "proxy/proxy-script.js";
+// Set the default list of domains to proxy on installation.
 
-// Default settings. If there is nothing in storage, use these values.
-const defaultSettings = {
-   blockedHosts: ["example.com", "example.org"]
- }
+// Listen for the extension's installation
+browser.runtime.onInstalled.addListener(handleInstalled);
 
-// Register the proxy script
-browser.proxy.register(proxyScriptURL);
+// On the extension's installation
+function handleInstalled(details) {
+// Write a blocked hosts list to local storage
+   browser.storage.local.set({
+     blockedHosts: ["example.com", "example.org"]
+   });
+}
+
+// Managed the proxy
+
+// Listen for a request to open a webpage
+browser.proxy.onRequest.addListener(handleProxyRequest, {urls: ["<all_urls>"]});
+
+// On the request to open a webpage
+function handleProxyRequest(requestInfo) {
+// Read the list of domains to be proxied
+  browser.storage.local.get().then(getList, onError);
+// Read the web address of the page to be visited 
+  const url = new URL(requestInfo.url);
+// Determine whether the domain in the web address is on the blocked hosts list
+  if (blockedHosts.indexOf(url.hostname) != -1) {
```

Comment:
This code relies on the existence of the `blockedHosts` variable, which doesn't exist until line 37 in `getList()`, where because it's declared without a keyword it's made global, so it is accessible in this function.

But, `getList()` is called from this function only after the promise returned by `storage.local.get()` resolves. So the first time this is run, `blockedHosts` is undefined at line 26 and the proxying won't happen. Only after that first time is `blockedHosts` going to contain correct values (and it will again be wrong for the first call after the list in local storage is changed).

You can see this by:
* installing the extension
* loading example.com - the page will load although it shouldn't
* reloading example.com - the page will be blocked, because `blockedHosts` was initialised by this time

It would be better for the call to get the list of blocked hosts to use async/await, so it gets the right values.

But still this is inefficient because you're reading local storage for every web request. So it might be better still to do what the previous version of this does, which is to keep a variable for the list of blocked hosts and only update it when local storage changes.

Reply:
Thanks for the feedback @chrisdavidmills the latest update should address it.